nni/retiarii/nn/pytorch/api.py

```diff
@@ -0,0 +1,289 @@
+                delattr(self, key)
+        else:
+            if isinstance(idx, str):
+                key, idx = idx, self.names.index(idx)
+            else:
+                key = self.names[idx]
+            delattr(self, key)
+        del self.names[idx]
+
+    def __len__(self):
+        return len(self.names)
+
+    def __iter__(self):
+        return map(lambda name: self._modules[name], self.names)
+
+    @property
+    def choices(self):
+        return self._choices()
+
+    @torch.jit.ignore
+    def _choices(self):
+        warnings.warn("layer_choice.choices is deprecated. Use `list(layer_choice)` instead.", category=DeprecationWarning)
+        return list(self)
+
+    def forward(self, x):
+        warnings.warn('You should be run forward of this module directly.')
+        return x
+
+
+class InputChoice(nn.Module):
+    """
+    Input choice selects ``n_chosen`` inputs from ``choose_from`` (contains ``n_candidates`` keys).
+    Use ``reduction`` to specify how chosen inputs are reduced into one output. A few options are:
+
+    * ``none``: do nothing and return the list directly.
+    * ``sum``: summing all the chosen inputs.
+    * ``mean``: taking the average of all chosen inputs.
+    * ``concat``: concatenate all chosen inputs at dimension 1.
+
+    We don't support customizing reduction yet.
+
+    Parameters
+    ----------
+    n_candidates : int
+        Number of inputs to choose from. It is required.
+    n_chosen : int
+        Recommended inputs to choose. If None, mutator is instructed to select any.
+    reduction : str
+        ``mean``, ``concat``, ``sum`` or ``none``. See :class:`LayerChoice`.
+    label : str
+        Identifier of the input choice.
+    """
+
+    def __init__(self, n_candidates: int, n_chosen: int = 1, reduction: str = 'sum', label: str = None, **kwargs):
+        super(InputChoice, self).__init__()
+        if 'key' in kwargs:
+            warnings.warn(f'"key" is deprecated. Assuming label.')
+            label = kwargs['key']
+        if 'return_mask' in kwargs:
+            warnings.warn(f'"return_mask" is deprecated. Ignoring...')
+        if 'choose_from' in kwargs:
+            warnings.warn(f'"reduction" is deprecated. Ignoring...')
+        self.n_candidates = n_candidates
+        self.n_chosen = n_chosen
+        self.reduction = reduction
+        self._label = label if label is not None else f'inputchoice_{uid()}'
+
+    @property
+    def key(self):
+        return self._key()
+
+    @torch.jit.ignore
+    def _key(self):
+        warnings.warn('Using key to access the identifier of InputChoice is deprecated. Please use label instead.',
+                      category=DeprecationWarning)
+        return self._label
+
+    @property
+    def label(self):
+        return self._label
+
+    def forward(self, candidate_inputs: List[torch.Tensor]) -> torch.Tensor:
+        warnings.warn('You should be run forward of this module directly.')
+        return candidate_inputs[0]
```

Comment:
please double check whether this works in our examples in test

Reply:
Checked.